src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala

```diff
@@ -362,12 +602,481 @@ object FunctionRegistry {
+    })
+
+  registerAggregator[Double, Double, Double, Int, Any]("hist", (start: Double, end: Double, bins: Int) => {
+    if (bins <= 0)
+      fatal(s"""method `hist' expects `bins' argument to be > 0, but got $bins""")
+
+    val binSize = (end - start) / bins
+    if (binSize <= 0)
+      fatal(
+        s"""invalid bin size from given arguments (start = $start, end = $end, bins = $bins)
+           |  Method requires positive bin size [(end - start) / bins], but got ${ binSize.formatted("%.2f") }
+                  """.stripMargin)
+
+    val indices = Array.tabulate(bins + 1)(i => start + i * binSize)
+
+    new HistAggregator(indices)
+  })(aggregableHr(doubleHr), doubleHr, doubleHr, intHr, new HailRep[Any] {
+    def typ = HistogramCombiner.schema
+  })
+
+  registerLambdaAggregator[Genotype, (Any) => Any, Any]("callStats", (vf: (Any) => Any) => new CallStatsAggregator(vf))(
+    aggregableHr(genotypeHr), unaryHr(genotypeHr, variantHr), new HailRep[Any] {
+      def typ = CallStats.schema
+    })
+
+  registerLambdaAggregator[Genotype, (Any) => Any, Any]("inbreeding", (af: (Any) => Any) => new InbreedingAggregator(af))(
+    aggregableHr(genotypeHr), unaryHr(genotypeHr, doubleHr), new HailRep[Any] {
+      def typ = InbreedingCombiner.signature
+    })
+
+  registerLambdaAggregator[Any, (Any) => Any, Any]("fraction", (f: (Any) => Any) => new FractionAggregator(f))(
+    aggregableHr(TTHr), unaryHr(TTHr, boxedboolHr), boxeddoubleHr)
+
+  registerAggregator("take", (n: Int) => new TakeAggregator(n))(
+    aggregableHr(TTHr), intHr, arrayHr(TTHr))
+
+  registerLambdaAggregator("takeBy", (f: (Any) => Any, n: Int) => new TakeByAggregator[Int](f, n))(
+    aggregableHr(TTHr), unaryHr(TTHr, boxedintHr), intHr, arrayHr(TTHr))
+  registerLambdaAggregator("takeBy", (f: (Any) => Any, n: Int) => new TakeByAggregator[Long](f, n))(
+    aggregableHr(TTHr), unaryHr(TTHr, boxedlongHr), intHr, arrayHr(TTHr))
+  registerLambdaAggregator("takeBy", (f: (Any) => Any, n: Int) => new TakeByAggregator[Float](f, n))(
+    aggregableHr(TTHr), unaryHr(TTHr, boxedfloatHr), intHr, arrayHr(TTHr))
+  registerLambdaAggregator("takeBy", (f: (Any) => Any, n: Int) => new TakeByAggregator[Double](f, n))(
+    aggregableHr(TTHr), unaryHr(TTHr, boxeddoubleHr), intHr, arrayHr(TTHr))
+  registerLambdaAggregator("takeBy", (f: (Any) => Any, n: Int) => new TakeByAggregator[String](f, n))(
+    aggregableHr(TTHr), unaryHr(TTHr, stringHr), intHr, arrayHr(TTHr))
+
+  val aggST = Box[SymbolTable]()
+
+  registerLambdaSpecial("filter", { (a: () => Any, f: (Any) => Any) =>
+    val x = a()
+    val r = f(x)
+    if (r != null && r.asInstanceOf[Boolean])
+      x
+    else
+      null
+  })(aggregableHr(TTHr, aggST), unaryHr(TTHr, boolHr), aggregableHr(TTHr, aggST))
+
+  registerLambdaSpecial("map", { (a: () => Any, f: (Any) => Any) =>
+    f(a())
+  })(aggregableHr(TTHr, aggST), unaryHr(TTHr, TUHr), aggregableHr(TUHr, aggST))
```

Comment:
I think this is hard to do with the current organization.  You really want aggregators to be pushed-based, rather than pull-based.  That is, instead of having `gs.filter(...).map(...).sum` evaluate everything to the left to get the next value, you really want `gs` to push values down the chain, have them short-circuit on filtering, iterate on flatMap, and hit the sum aggregator at the end.

We should move towards this but I felt it too much for this change.